osquery/tables/system/windows/groups.cpp

```diff
@@ -0,0 +1,142 @@
+  auto eSidType = SidTypeUnknown;
+  LookupAccountNameW(nullptr,
+                     accountName,
+                     nullptr,
+                     &sidBufferSize,
+                     nullptr,
+                     &domainNameSize,
+                     &eSidType);
+
+  // Allocate sufficient buffers for the (binary data) SID and the (wide string)
+  // domain name:
+  auto sidBuffer = std::make_unique<BYTE[]>(sidBufferSize);
+  std::vector<wchar_t> domainName(domainNameSize);
+
+  // Call LookupAccountNameW() a second time to actually obtain the SID for the
+  // given account name:
+  auto ret = LookupAccountNameW(nullptr,
+                                accountName,
+                                sidBuffer.get(),
+                                &sidBufferSize,
+                                domainName.data(),
+                                &domainNameSize,
+                                &eSidType);
+  if (ret == 0) {
+    LOG(INFO) << "Failed to LookupAccountNameW(): " << accountName;
+  } else if (IsValidSid(sidBuffer.get()) == FALSE) {
+    LOG(INFO) << "The SID for " << accountName << " is invalid.";
+  }
+
+  return sidBuffer; // Implicit move operation. Caller "owns" returned pointer.
+}
+
+void processLocalGroups(QueryData& results) {
+  unsigned long groupInfoLevel = 1;
+  unsigned long numGroupsRead = 0;
+  unsigned long totalGroups = 0;
+  unsigned long resumeHandle = 0;
+  unsigned long ret = 0;
+  LOCALGROUP_INFO_1* lginfo = nullptr;
+
+  std::unique_ptr<BYTE[]> sidSmartPtr = nullptr;
+  PSID sidPtr = nullptr;
+
+  do {
+    ret = NetLocalGroupEnum(nullptr,
+                            groupInfoLevel,
+                            (LPBYTE*)&lginfo,
+                            MAX_PREFERRED_LENGTH,
+                            &numGroupsRead,
+                            &totalGroups,
+                            nullptr);
+
+    if ((ret == NERR_Success || ret == ERROR_MORE_DATA) && lginfo != nullptr) {
+      for (size_t i = 0; i < numGroupsRead; i++) {
+        Row r;
+        sidSmartPtr = GetSid(lginfo[i].lgrpi1_name);
+
+        if (sidSmartPtr != nullptr) {
+          sidPtr = static_cast<PSID>(sidSmartPtr.get());
+        } else {
+          // nullptr still valid, just results in blank fields
+          sidPtr = nullptr;
+        }
+
+        // Windows' extended schema, including full SID and comment strings:
+        r["group_sid"] = psidToString(sidPtr);
+        r["comment"] = wstringToString(lginfo[i].lgrpi1_comment);
+
+        // Common schema, normalizing group information with POSIX:
+        r["gid"] = INTEGER(getRidFromSid(sidPtr));
+        r["gid_signed"] = INTEGER(getRidFromSid(sidPtr));
+        r["groupname"] = wstringToString(lginfo[i].lgrpi1_name);
+        results.push_back(r);
+      }
+    } else {
+      LOG(INFO) << "NetLocalGroupEnum failed with return value: " << ret;
+    }
+    if (lginfo != nullptr) {
+      NetApiBufferFree(lginfo);
+    }
+  } while (ret == ERROR_MORE_DATA);

```

Comment:
Are you sure a `do/while` is what you want here? Looks like if the `NetLocalGroupEnum` function returns `ERROR_MORE_DATA` you just need to increase the size of the buffer you're passing. Perhaps this loop could be removed in favor of a couple of subsequent calls to the API to get a proper sized buffer?

Reply:
As I did in `getSid()`, yes, it seems like you should be able to call the API once to get a size and a second time to get the result. But, as I think about it, this may risk a race condition bug, in a hypothetical case where the OS increased the size of its objects during the time it took for you to allocate a buffer. I.e. "how big is this data?" "It's 100 bytes" "Ok, I got a 100 byte buffer ready" "Oh it's 110 bytes now, ERROR_MORE_DATA".

Every Windows API example from Microsoft uses a loop to call around these type of APIs that fill a user-supplied buffer argument. That makes it pretty standard practice, but, maybe you don't like the potential of an infinite loop in osquery code?